registry/remote_factory.go

```diff
@@ -0,0 +1,131 @@
+// We can't inject a remote client directly because each repository request might be to a different
+// registry provider. E.g. both docker hub and quay containers. So a new remote client must be
+// created for each new image. This factory provides that and can be mocked out.
+package registry
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	dockerregistry "github.com/heroku/docker-registry-client/registry"
+	"github.com/jonboulle/clockwork"
+	"github.com/weaveworks/flux"
+	"golang.org/x/net/publicsuffix"
+	"net/http"
+	"net/http/cookiejar"
+	"strings"
+)
+
+type creds struct {
+	username, password string
+}
+
+// Credentials to a (Docker) registry.
+type Credentials struct {
+	m map[string]creds
+}
+
+type RemoteClientFactory interface {
+	Create(id Image) (Remote, error)
+}
```

Comment:
Ok, so above you've extracted out the creds. That's cool. But you've effectively just renamed the factory method to resolve. Since you have an aversion to the word factory, I will blast all mentions off the face of the codebase. :-)